src/main/groovy/de/dkfz/roddy/bam2fastq/BamToFastqWorkflow.groovy

```diff
@@ -0,0 +1,266 @@
+        }
+    }
+
+    protected FileGroup cleanupFastqsForBam(Config config, BaseFile controlBam, List<String> readGroups) {
+        if (config.pairedEnd) {
+            List<String> rgFileIndicesParameter = readGroupIndices(readGroups)
+            return callWithOutputFileGroup(TOOL_CLEANUP, controlBam, rgFileIndicesParameter, [readGroups: readGroups])
+        } else {
+            throw new ConfigurationError("Single-end cleanup not implemented", config.FLAG_PAIRED_END)
+        }
+    }
+
+    @Override
+    boolean cleanup() {
+        for (BaseFile BaseFile : _bamFilesPerDataset.get(context.dataSet, [])) {
+
+            if (config.outputPerReadGroup) {
+                cleanupFastqsForBam(config, BaseFile, readGroupsPerBamfile[BaseFile.getAbsolutePath()])
+            }
+
+        }
+        return true
+    }
+
+    Map<String, List<String>> readAllReadGroups(List<BaseFile> bamFiles) {
+        return bamFiles.collectEntries { bamFile ->
+            new MapEntry(bamFile.absolutePath, listReadGroups(bamFile.absolutePath))
+        }
+    }
+
+    @Override
+    boolean setupExecution(ExecutionContext context) {
+        _config[context.dataSet] = new Config(context)
+
+        boolean result = super.setupExecution(context)
+        determineBamFiles()
+        if (bamFiles.size() > 0) {
+            if (new Config(context).outputPerReadGroup)
+                readAllReadGroups(bamFiles)
+            result &= true
+        } else {
+            result &= false
+        }
+        return result
+    }
+
+    @Override
+    boolean execute(ExecutionContext context) {
+
+        for (BaseFile bamFile : bamFiles) {
+
+            if (config.outputPerReadGroup) {
+                List<String> readGroups = readGroupsPerBamfile[bamFile.absolutePath]
+
+                FileGroup unsortedFastqs = bam2fastq(bamFile, readGroups)
+
+                if (config.sortFastqs) {
+                    for (int i = 0; i < readGroups.size(); i++) {
+                        this.sortFastqs(bamFile.getPath().name, readGroups[i], unsortedFastqs[i * 2], unsortedFastqs[i * 2 + 1]) // All paired end, sorted by read group.
+                    }
+                }
+
+            } else {
+//                FileGroup out = bam2fastq(cfg, BaseFile)
+//                if (cfg.sortFastqs())
+//                    this.sortFastqs(cfg)
+                throw new ConfigurationError("bam2fastq without output per read group not implemented", config.FLAG_SPLIT_BY_READ_GROUP)
+            }
+
+        }
+
+        return true
+    }
+
+    protected boolean bamFilesAreAccessible(List<BaseFile> bamFiles) {
+        return bamFiles.collect { file ->
+            boolean result = context.fileIsAccessible(file.path)
+            if (!result)
+                context.addErrorEntry(ExecutionContextError.EXECUTION_PATH_INACCESSIBLE.
+                        expand("BAM file not accessible: '${file.absolutePath}'"))
+            result
+        }.inject { res, i -> res && i }
```

Comment:
I don't get it... what does it? Maybe a simple comment, thanks for explaining it.

Reply:
Just found out, it is .every() what you need.